Refactor Wordpress controller

muvaf opened the pull request:
Refactor Wordpress controller:
- Use typed resources instead of raw YAML and templating.
- Spec.Image is introduced to be able to change the Wordpress image to deploy.
- Status.Endpoint is introduced to let user access the endpoint of Wordpress
instance from WordpressInstance custom resource instead of finding service manually.

I might have gone too far on this. I only wanted to convert it to use typed resources but without good separation, it just looked like a giant hack function with a lot of nested objects. One thing led to another and here we are. Please don't hesitate to suggest fundamental changes.

FIxes #16

muvaf commented:
FYI, I still need to test this PR manually.

suskin commented on api/v1alpha1/wordpressinstance_types.go:
```diff
@@ -28,12 +28,21 @@ import (
+	// Endpoint is the URL of Wordpress that you can use to access.
+	// It will be populated as soon as a LoadBalancer is assigned to Wordpress
+	// service.
+	Endpoint string `json:"endpoint,omitempty"`
```
Yesss

suskin commented on controllers/wordpressinstance_controller.go:
```diff
@@ -50,283 +59,90 @@ type WordpressInstanceReconciler struct {
-			return err
-		}
+func ApplyResources(ctx context.Context, kube client.Client, wp wordpressv1alpha1.WordpressInstance) error {
+	database := GetMySQLInstance(wp)
```
I believe it may be better if all of the resources were applied first, then all errors were collected, and then all errors were returned. That way, resources without errors which may take time to spin up (such as the cluster heh heh) can still make progress even if something else errors out. I believe it'll also make the code a little more friendly since all of the resources will be declared right next to each other in the code.

muvaf replied:
Controller-runtime supports client operations only on 1 resource at a time and AFAIK checking the error right away is what I see as a usual pattern in Go. The error accumulation usually done when you spawn more than one goroutines and need to check their errors.

I've tried a few things to do what you suggest anyway but they all looked like hack to be honest. Creating a list, calling the operation on each one of them, checking the list of errors and returning a meaningful error didn't seem worth it. In fact, most of the time creation doesn't fail when the resources are strongly-typed(validation passes in almost all cases). I'd expect the errors that you refer to to appear because of provisioning problems with cloud providers, which the controller doesn't really look at right now.

suskin commented on controllers/wordpressinstance_controller.go:
```diff
@@ -50,283 +59,90 @@ type WordpressInstanceReconciler struct {
-	// if we decoded a non-nil unstructured object, try to create it now
-	if obj == nil {
-		return nil
+func GetEndpoint(ctx context.Context, kube client.Client, wp wordpressv1alpha1.WordpressInstance) (string, error) {
```
I'm super happy about bubbling up the endpoint into the instance - nice!

suskin commented on controllers/wordpressinstance_controller.go:
```diff
@@ -50,283 +59,90 @@ type WordpressInstanceReconciler struct {
-		Namespace:  o.GetNamespace(),
-		Name:       o.GetName(),
-		UID:        o.GetUID(),
+func FillWithDefaults(wp *wordpressv1alpha1.WordpressInstance) {
```
I'm not super up on the go patterns for doing this, but I would normally prefer to see a default-setting method as taking the defaults as a parameter, then merging the defaults into the object. The behavior would be similar to a deep copy into an existing object. Is that something that is easy and idiomatic in go? One nice benefit is that the function is stateless if the default object is passed in. Another nice benefit is that it's easy to add more defaults by just adding them to the default object.

muvaf replied:
I am not sure I capture what you mean by default object. Is it something like the following;
```
var default = WordpressInstance{... default params ...}
func FillWithDefaults(wp *wordpressv1alpha1.WordpressInstance){
  if wp.Spec.Image == "" {
    wp.Spec.Image = default.Spec.Image 
  }
  ... repeate if block for each statement...
}
```
If so, adding additional defaults is not much different than current case actually.

suskin replied:
Yeah, similar to that. I'm imagining something more like a stateless object-merging function like [this one in JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign):

```
var default = WordpressInstance{... default params ...}
func FillWithDefaults(wp *wordpressv1alpha1.WordpressInstance, defaultwp *wordpressv1alpha1.WordpressInstance){
  if wp.Spec.Image == "" {
    wp.Spec.Image = defaultwp.Spec.Image 
  }
  ... repeat if block for each statement...
}
```

Then, the call passes in the defaults:
```
FillWithDefaults(wp, default)
```

Actually, now that I wrote this, I'm realizing, can you use `DeepCopyInto` instead of this filling logic? I'm seeing `DeepCopyInto` in the generated deep copy file.

```
var wp = // this is fetched from the client
var default = WordpressInstance{... default params ...}
var myWp = WordpressInstance{}

default.DeepCopyInto(&myWp)
wp.DeepCopyInto(&myWp)
```

muvaf replied:
> can you use DeepCopyInto instead of this filling logic?

`DeepCopyInto` assigns the values even if a value is already there. So, it wouldn't work for defaulting.

I am actually leaning towards keeping it as is since it's only 1 field now and proposal doesn't seem like providing much value, i.e. still you add a statement when a new field with a default arrives.

suskin replied:
> DeepCopyInto assigns the values even if a value is already there. So, it wouldn't work for defaulting.

Does it assign the values if the `from` object has empty values? It [seems like it does not](https://github.com/crossplaneio/sample-stack-wordpress/pull/17/files#diff-a9ceb1ef9a800df85f5f80f38e4b6deeR89)

I was talking about the approach of starting with a third empty object. Then layer the default fields in, then layer the original object's fields in. Here is a more complete example of what I was imagining:

```go
// defaults for a wordpress instance
const defaultWp = WordpressInstance{... default params ...}

var clientWp = // this is fetched from the client.Get call
var wpInstance = FillWithDefaults(&clientWp, &defaultWp)

// then, use wpInstance
// ...

// meanwhile . . .
func FillWithDefaults(wp *wordpressv1alpha1.WordpressInstance, defaultWp *wordpressv1alpha1.WordpressInstance)  wordpressv1alpha1.WordpressInstance {
  // this starts off empty, but later stores the final merged result
  var resultWp = WordpressInstance{}

  defaultWp.DeepCopyInto(&resultWp)
  wp.DeepCopyInto(&resultWp)

  return resultWp
}
```

muvaf replied:
> Does it assign the values if the from object has empty values? It seems like it does not

I think this is correct only for pointer types though. Non-pointers are [directly copied](https://github.com/crossplaneio/sample-stack-wordpress/pull/17/files#diff-a9ceb1ef9a800df85f5f80f38e4b6deeR88)

I just tested it and indeed the `Image string` field gets overridden with empty string from the actual `WordpressInstance` object.

suskin replied:
Dang! Ok. Thanks for testing it!

suskin commented on controllers/resources.go:
```diff
@@ -0,0 +1,230 @@
+			Labels: labelsInRemote,
+		},
+	})
+	localDeploymentResourceName := fmt.Sprintf("wordpress-demo-deployment-%s-%s", wp.GetNamespace(), wp.GetName())
```
Now that we've moved beyond a heredoc with a template, I am not a huge fan of the hard-coded string prefixes, since they're a bit harder to work with if we ever want to change them . . . what are your thoughts on pulling them up into constants somewhere so they're easier to change?

suskin commented on controllers/wordpressinstance_controller.go:
```diff
@@ -50,283 +59,90 @@ type WordpressInstanceReconciler struct {
-		"namespace": instanceNamespace,
+	FillWithDefaults(wp)
+	if err := ApplyResources(ctx, r.Client, *wp); err != nil {
+		return ctrl.Result{RequeueAfter: shortWait}, err
```
Yay we have error handling now!

negz commented:
> One thing led to another and here we are. 

Haha, this is representative of like 90% of my PRs.

suskin commented on controllers/wordpressinstance_controller.go:
```diff
@@ -50,283 +59,90 @@ type WordpressInstanceReconciler struct {
+	if err := kube.Patch(ctx, &cluster, client.Apply); err != nil && !kerrors.IsAlreadyExists(err) {
+		return errors.Wrap(err, fmt.Sprintf("cannot create %s/%s in namespace %s", cluster.Kind, cluster.GetName(), cluster.GetNamespace()))
+	}
+	kubernetesApplication := GetKubernetesApplication(wp, database.GetName())
```
I like the general direction we're going of making things nicer and easier to work with. I'm trying to think about how we can also make the code as easy to follow as possible; yaml may not be very nice to work with, but it's pretty easy to follow the flow of (since there is no flow, haha).

Anyway, one adjustment which I think could make sense is to the way that the names are passed around, so that the data only flows in one direction. I would recommend that we create the names *outside* of resources.go, at the top of this function. Then we can inject them into all of the templating-equivalent methods. Specifically, I feel that the way we are currently creating the database name in `GetMySQLInstance`, then returning an object, then threading that object's name into `GetKubernetesApplication`, is a bit hard to follow at first. But it could just be me. What do you think? My real goal is to make reading the code as simple as possible, so it could be the case that doing this would also work just as well:

```go
database := GetMySQLInstance(wp)
cluster := GetKubernetesCluster(wp)
kubernetesApplication := GetKubernetesApplication(wp, database.GetName())

// then, call patch a bunch of times and collect the errors
// then, return
```

muvaf replied:
For batching, see https://github.com/crossplaneio/sample-stack-wordpress/pull/17/files#r347071754

For name generations; one can have a `func GenerateMySQLInstanceName(v1alpha1.WordpressInstance)` and use this in both `GetKubernetesApplication` and `GetMySQLInstance(wp)`. But I like the explicitness that we give the name of the _produced_ database. At some point, `GetMySQLInstance` could be changed to use another name or another name generation function that is different than what's used in `GetKubernetesApplication`

suskin commented:
> > One thing led to another and here we are.
> 
> Haha, this is representative of like 90% of my PRs.

I also loled

muvaf commented:
@suskin I've tried addressing your comments. Please take a look at second look.

There are couple of things that could be done but I'm not sure we want to do it since this is just a sample stack that is supposed to be _simple_.

- Update image of Wordpress with value from `spec.image`.
- Detailed `status` that has more than just `endpoint`. 
- More spec fields that are tied to Wordpress.
- More spec fields for K8S and MySQL configurations.
- Update created claims.

I mean I can go really deep and create a very robust controller but it's just I'm not sure how simple the code should be for this repo.

muvaf commented:
And +1 for template stack controller work that is going on. I can see that it will make writing app stacks much much easier!

suskin commented on controllers/wordpressinstance_controller.go:
```diff
@@ -50,283 +64,153 @@ type WordpressInstanceReconciler struct {
-			r.Log.V(2).Info("EXIT extract objects because ERROR", "objects", objects, "err", err)
-			return nil, errors.Wrapf(err, "failed to parse output")
+func CreateMySQLInstance(ctx context.Context, kube client.Client, wp *wordpressv1alpha1.WordpressInstance) error {
+	if wp.Spec.MySQLInstanceRef != nil {
```
Now that our code is starting to look more "realistic", could we add some comments where we diverge from Best Practices or have gotchas, and explain what the best practices would be or what the gotchas are?

For example, here, we may create multiple databases if multiple workers see a missing ref at the same time before the reference is recorded. With a production-grade controller, it may make sense to have some code for mitigating that scenario.

Another example would be that our code appears to be updating the fetched wordpressinstance object directly, but that [isn't recommended](https://github.com/kubernetes/sample-controller/blob/65e4a1173959df6b96adf627c491057feeb67e19/controller.go#L321). Or maybe the comment should be that we're fetching the object directly from the API, but you may want to use a local lister cache instead?

Anyway, comments like that.

muvaf replied:
> we may create multiple databases if multiple workers see a missing ref at the same time before the reference is recorded.

Actually, there is always 1 worker going through the `Reconcile` at any time. Also, since the name is static, once a creation succeeds, other creation operations would fail and the case where the resource already exists is handled in the `Create` call. Unless I'm missing something, this function looks idempotent to me.

> Another example would be that our code appears to be updating the fetched wordpressinstance object directly, but that isn't recommended.

That recommendation is valid if you use go-client directly though. controller-runtime _deep-copies_ into the object you yourself created and provided in the [`Get` call.](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/typed_client.go#L145)

suskin replied:
> controller-runtime deep-copies

Ahh cool

suskin commented on controllers/wordpressinstance_controller.go:
```diff
@@ -50,283 +64,153 @@ type WordpressInstanceReconciler struct {
+	if err := kube.Get(ctx, key, serviceResource); err != nil {
+		return "", err
+	}
+	if serviceResource.Status.Remote == nil {
```
What is this case?

muvaf replied:
There is a case where the service resource has `serviceResource.Status` but `serviceResource.Status.Remote` is actually nil. In the next statement, we access `serviceResource.Status.Remote.Raw` which would cause panic when `serviceResource.Status.Remote` is nil.

suskin commented on controllers/wordpressinstance_controller.go:
```diff
@@ -50,283 +64,153 @@ type WordpressInstanceReconciler struct {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&wordpressv1alpha1.WordpressInstance{}).
-		Complete(r)
+func GetEndpoint(ctx context.Context, kube client.Client, wp wordpressv1alpha1.WordpressInstance) (string, error) {
```
This seems like a lot of boilerplate, so maybe it'd make sense to add a comment explaining why this method and all of its logic exists

suskin commented on controllers/resources.go:
```diff
@@ -0,0 +1,283 @@
+	}
+}
+
+func ConvertToUnstructured(o metav1.Object, scheme *runtime.Scheme) (*unstructured.Unstructured, error) {
```
What is the reason this exists? It seems like we may want a comment explaining the reasons, for future reference

muvaf replied:
Because `KubernetesApplication` type requires you to send your objects as `*unstructured.Unstructured` instead of directly `Deployment` or `Service`. This is why that function exists but I'm not sure whether I should write this as function's comment. Because you can just use it for some other reason. I'll add this reasoning where it's called for that purpose for clarity.

muvaf replied:
@suskin actually I'm not sure whether that comment would be useful. It's a classic type conversion like json marshal that's needed for the corresponding field's type.

suskin replied:
@muvaf it's okay if it can be used for a different reason; that comment is still useful. I would find it helpful as a future reader. The goal of adding a comment is to capture the reason the method was originally created, so a reader or future maintainer can get into the original author's mindset at the time the code was written.

muvaf replied:
Alright, makes sense. Just added the comment. Also realized using `metav1.Object` instead of `runtime.Object` may not make sense at the first glance, so, I added an explanation for that as well.

suskin replied:
Awesome. The `The reason metav1.Object is used . . .` is really helpful. For the other part, this part is a bit redundant, since it's in the method signature:

```go
// ConvertToUnstructured takes a Kubernetes object and converts it into
// *unstructured.Unstructured
```

I liked this part from your earlier explanation, so I think it'd make sense to have it in the comment:

> KubernetesApplication type requires you to send your objects as *unstructured.Unstructured instead of directly Deployment or Service

For example, something like:

```go
// ConvertToUnstructured exists because the KubernetesApplication type requires you 
// to send your objects as *unstructured.Unstructured instead of directly as Deployment or Service.
```

But, I don't think it's worth blocking the merge over; we can improve it in the future.

suskin commented:
@muvaf I've assigned this PR to you - feel free to review my clarification about DeepCopyInto, and then change or add comments and merge at your convenience after that

suskin commented:
> @muvaf I've assigned this PR to you - feel free to review my clarification about DeepCopyInto, and then change or add comments and merge at your convenience after that

@muvaf or, assign it back to me if you want me to give it another look before it is merged or if you want me to merge it

jbw976 commented:
@suskin have you already published an updated wordpress stack package? we don't have CI or automatic publishing on this repo, so hopefully you can publish a new one manually so the latest is this new fanciness from @muvaf 🚀
